R/fragments-calculate2.R

```diff
@@ -0,0 +1,287 @@
+        cmz_mod[[i]] <- rep(cmz_mod[[i]], nct) + rep(add[4:6][ctype], each=tn)
+        
+        amz_mod[[i]] <- rep(amz_mod[[i]], each = zn)/z
+        cmz_mod[[i]] <- rep(cmz_mod[[i]], each = zn)/z
+        
+        ## add protons (H+)
+        amz_mod[[i]] <- amz_mod[[i]] + mass["p"]
+        cmz_mod[[i]] <- cmz_mod[[i]] + mass["p"]
+    }
+    
+    ## fragment str (amino-terminus)
+    atype <- rep(c("a", "b", "c")[atype], each = tn * zn)
+    pos <- rep(1L:tn, each = zn)
+    if (length(atype)) {
+        aion <- paste0(atype, pos)
+    } else {
+        aion <- character()
+    }
+    
+    ## fragment str (carboxyl-terminus)
+    ctype <- rep(c("x", "y", "z")[ctype], each = tn * zn)
+    if (length(ctype)) {
+        cion <- paste0(ctype, pos)
+    } else {
+        cion <- character()
+    }
+    
+    ## generate unique dataframe with all fragments and modifications 
+    for (i in 1:length(mod_combinations)) {
+        df[[i]] <- data.frame(mz = c(amz_mod[[i]], cmz_mod[[i]]),
+                              ion = c(aion, cion),
+                              type = c(atype, ctype),
+                              pos = pos,
+                              z = z,
+                              seq = c(aseq, cseq),
+                              stringsAsFactors = FALSE)
+        df[[i]] <- .neutralLoss(df[[i]],
+                                water = neutralLoss$water,
+                                ammonia = neutralLoss$ammonia)
+        df[[i]] <- .terminalModifications(df[[i]],
+                                          modifications = fixed_modifications)
+        rownames(df[[i]]) <- NULL
+        non_zero <- mod_combinations[[i]] != 0
+        names(mod_combinations[[i]])[non_zero] <- 
+            paste0(names(mod_combinations[[i]])[non_zero],
+                   "[",
+                   mod_combinations[[i]][non_zero],
+                   "]")
+        df[[i]][["peptide"]] <- paste(names(mod_combinations[[i]]),
+                                      collapse = "")
+    }
+    
+    df <- do.call(rbind, df)
+    rownames(df) <- NULL
+    df
+}
+
+#' @title Generates list of possible combinations of modifications
+#' 
+#' @param sequence Character. A peptide sequence that may have modifications or not 
+#' 
+#' @param fixed_modifications Named numeric. Specifies which fixed modifications are used
+#' 
+#' @param variable_modifications Named numeric. Specifies which variable modifications are used
+#' 
+#' @param max_mods Numeric. Indicates how many modifications can be applied at once.
+#' 
+#' @return list with all possible combinations of modifications
+#' 
+#' @noRd
+#' 
+#' @examples
+#' .modificationPositions("ARGHKA", variable_modifications = c(A = 4, K = 5, S = 8), max_mods = 3)
+#' 
+.modificationPositions <- function(sequence,
+                                   variable_modifications = numeric(),
+                                   max_mods = Inf) {
+    sequence_split <- strsplit(sequence, "", fixed = TRUE)[[1]]
```

Comment:
This is just a minor point. We already split the sequence in `calculateFragments2`. So you maybe want to avoid this line by use `fragment.seq` as argument?